### ISSUE TITLE
Working local code does not work when deployed

### ISSUE THREAD

**Issue by prakashsanker** [serverless/serverless#4417]
# This is a Bug Report

## Description

I have a super weird bug and I think I've narrowed it down to it being an issue with either Serverless or Lambda. Either way, this is probably the best place to get help.

I basically have code that 100% works in these two scenarios

1. When I am using serverless-offline and connecting to a local Mongo DB.
2. When I am using serverless-offline and connection to a remote Mongo DB (Atlas).

However, when I deploy this code as a Lambda function it does not work.

I execute an this function from my endpoint

```
export async function updateUser(data, userData, token, connection) {
  userData = JSON.parse(userData);
  if (verifyToken(data, token)) {
    if (userData) {
      try {
        const userModel = new UserModel(connection);
        const jobModel = new JobModel(connection);
        try {
          let updatedUser = await userModel.update(data.userId, userData);
          updatedUser = updatedUser.toObject();
          const jobs = await jobModel.getJobs();
          // candidate for optimization.
          _.each(jobs, job => {
            algorithmUtils.filterScoreAndUpdateMatches(
              [updatedUser],
              job,
              connection
            );
          });
          return updatedUser;
        } catch (error) {
          throw error;
        }
      } catch (error) {
        throw error;
      }
    } else {
      throw new Error("Invalid user data");
    }
  } else {
    throw new Error("Authentication Failed");
  }
}
```

`algorithmUtils.filterScoreAndUpdateMatches` looks like this -

```
export async function filterScoreAndUpdateMatches(candidates, job, connection) {
  const filteredCandidates = filter(candidates, job);
  console.log("FILTERED CANDIDATES");
  console.log(filteredCandidates);
  const scoredCandidates = await score(filteredCandidates, job, connection);
  console.log("SCORED CANDIDATES");
  console.log(scoredCandidates);
  createOrUpdateMatches(scoredCandidates, job, connection);
}
```

And the relevant part of score looks like this -
```
export async function score(candidates, job, connection) {
  let scoredCandidates;
  const skillScoreModel = new SkillScoreModel(connection);
  const roleScoreModel = new RoleScoreModel(connection);
  const industryScoreModel = new IndustryScoreModel(connection);
  const jobFunctionModel = new JobFunctionModel(connection);
  const testModel = new TestModel(connection);
  const skillModel = new SkillModel(connection);
  console.log("JOB");
  console.log(job);
  let jobFunction;
   jobFunction = await jobFunctionModel.getJobFunction(job.function);
  console.log("JOB FUNCTION");
  console.log(jobFunction);
```

When looking at Cloudwatch, the `JOB FUNCTION` console.log never runs.

Looking at `JobFunctionModel` 

```

export default class JobFunctionModel extends BaseModel {
  constructor(connection) {
    super("JobFunction", connection);
    this.name = "JobFunction";
    this.schema = jobFunctionSchema;
    this.connection = connection;
    try {
      this.model = this.connection.model(this.name, this.schema);
    } catch (e) {
      console.log(e);
    }
  }

  async getJobFunctions() {
    if (!this.model) {
      await this._getModel();
    }

    try {
      const jobFunctions = await this.model.find({});
      return jobFunctions;
    } catch (e) {
      throw e;
    }
  }

  async getJobFunction(jobFunctionId) {
    if (!this.model) {
      await this._getModel();
    }
    try {
      const jobFunction = await this.model.findById(jobFunctionId); 
      console.log("DOES THIS RUN?");
      return jobFunction;
    } catch (e) {
      throw e;
    }
  }

  async getWeightsForJobFunction(functionDescription) {
    if (!this.model) {
      await this._getModel();
    }

    try {
      const weights = await this.model.find({
        description: functionDescription
      });
      return weights;
    } catch (e) {
      throw e;
    }
  }
}
```
The failing line is the one with the call to `findById`. Putting a console log after that line - that never runs. It's not just this command - when I run ANY query with the model, nothing else runs and there is no error.

I've tried

1. Looking at the connection to see if it was still alive (it is).
2. Tried different commands (none work).
3. console logged the connection object (it has all of the correct schemas)
4. Putting a try catch around the db command to see if there is an error - nothing.

There is *no* difference between my local code and my staging code, yet local runs fine, but staging just doesn't seem to run after I use the job function model.

Serverless Framework Version - 1.17.0
Operating System - Mac OS Sierra 10.12.1

Cloud front logs -

These correspond to code that looks like this -

```
  async getJobFunction(jobFunctionId) {
    if (!this.model) {
      await this._getModel();
    }
    console.log("IN GET JOB FUNCTION");
    console.log(jobFunctionId);
    console.log("MODEL");
    console.log(this.model);
    console.log("CONNECTION");
    console.log(this.connection);
    const test = await this.model.find({});
    console.log("TEST");
    console.log(test);
    console.log(this.model);
    try {
      const jobFunction = await this.model.findById({ _id: jobFunctionId });
      return jobFunction;
    } catch (e) {
      throw e;
    }
  }
```

The logs -
```
00:00:40
START RequestId: b3badccb-b9e0-11e7-8fe9-d32e065b1aa4 Version: $LATEST
2017-10-26T00:00:45.115Z	b3badccb-b9e0-11e7-8fe9-d32e065b1aa4	IN GET JOB FUNCTION
00:00:45
2017-10-26T00:00:45.115Z	b3badccb-b9e0-11e7-8fe9-d32e065b1aa4	d1809b80-b8fe-11e7-ba46-69c0134cb5f3
00:00:45
2017-10-26T00:00:45.115Z	b3badccb-b9e0-11e7-8fe9-d32e065b1aa4	MODEL
00:00:45
2017-10-26T00:00:45.115Z	b3badccb-b9e0-11e7-8fe9-d32e065b1aa4	{ [Function: model] hooks: Kareem { _pres: {}, _posts: {} }, base: Mongoose { connections: [ [Object], [Object] ], models: {}, modelSchemas: { User: [Object], Test: [Object], Question: [Object], Response: [Object], Industry: [Object], Role: [Object], Skill: [O

00:00:45
2017-10-26T00:00:45.241Z	b3badccb-b9e0-11e7-8fe9-d32e065b1aa4	CONNECTION
00:00:45
2017-10-26T00:00:45.241Z	b3badccb-b9e0-11e7-8fe9-d32e065b1aa4	NativeConnection { base: Mongoose { connections: [ [Object], [Circular] ], models: {}, modelSchemas: { User: [Object], Test: [Object], Question: [Object], Response: [Object], Industry: [Object], Role: [Object], Skill: [Object], Job: [Object], Ski
00:00:45
END RequestId: b3badccb-b9e0-11e7-8fe9-d32e065b1aa4
00:00:45
REPORT RequestId: b3badccb-b9e0-11e7-8fe9-d32e065b1aa4	Duration: 4411.14 ms	Billed Duration: 4500 ms Memory Size: 1024 MB	Max Memory Used: 68 MB
```

The relevant part of my `serverless.yml`

```
updateUser:
    handler: src/endpoints/user.put
    events:
      - http:
          path: /users/{userId}
          method: put
          cors:
            origin: "*"
            headers:
              - Content-Type
              - X-Amz-Date
              - Authorization
              - X-Api-Key
              - X-Amz-Security-Token
              - X-Amz-User-Agent
              - some_custom_header
              - Access-Control-Allow-Headers
            allowCredentials: false
```

I also set `context.callbackWaitsForEmptyEventLoop = false;` in the endpoint and cache the mongoose connection.

What can I do? How can I diagnose this?

**Comment by horike37** [comment]
Thank you for reporting @prakashsanker :+1: 
I think the code has a root cause. Local invoke can't perfectly emulate Lambda, so the result by Lambda is true rather than local one.

On the other hand, Serverless would not have a root cause since it is just deploy-tool and don't affect your code in most cases. To confirm this, you can use `serverless package` and generate CloudFormation template for deployment, then directly deploy that via CloudFormation on AWS console.

**Comment by prakashsanker** [comment]
I don't think the problem is in my code as it runs fine locally and what you say makes sense. I guess the issue is likely with Lambda?

**Comment by prakashsanker** [comment]
Ok I figured this out. I think it's worth my writing an explanation in case anybody else runs into this.

It comes down to this

```
        _.each(jobs, job => {
          algorithmUtils.filterScoreAndUpdateMatches(
            [createdUser],
            job,
            connection
          );
        });
```

`filterScoreAndUpdateMatches` is async, but I figured I didn't have to to `await` on it because the return value of the function didn't rely on this. This function works almost like a job that's kicked off. 

Basically, I was assuming that the lambda function would exist _until_ all functions that it kicked off finished executing. That's not the case, it looks like the lambda function stops existing as soon as the endpoint returns (or some time after that). 

Instead I had to do 

```
          const algoPromises = _.map(jobs, async job => {
            return await algorithmUtils.filterScoreAndUpdateMatches(
              [updatedUser],
              job,
              connection
            );
          });
        await algoPromises;
```

Now clearly there's a perf hit - I have to wait for the function to return (and it executes a bunch of queries). I imagine the solution (once I see that the queries are becoming too much) is to delegate this to another microservice (serverless or with a server).
